<a href="https://colab.research.google.com/github/SaiMudigonda/Saikumar_INFO5731_Spring2021/blob/main/In_class_exercise_08_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **The eighth in-class-exercise (20 points in total, 3/30/2021)**

The data for this exercise is from the dataset you created from assignment three. Please perform answer the following questions based on your data:

## (1) (10 points) Write a python program to extract the sentiment related terms from the corpus. You may use python package such as polyglot or external lexicon resources in the question. Rank the sentiment related terms by frequency.

In [3]:
# Pre-processing
from nltk.corpus import stopwords
import nltk
nltk.download('stopwords')
import re
from textblob import TextBlob
from textblob import Word
nltk.download('wordnet')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [4]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/SaiMudigonda/Saikumar_INFO5731_Spring2021/main/Reviews%20for%20Sentiment%20Analysis.csv', encoding='latin-1')
df

Document_ID                                       Cleaned Data Sentiment
0      Review 1  Every once in a while a movie comes, that trul...  Positive
1      Review 2  This is a movie that only those who have felt ...  Positive
2      Review 3  Truly a masterpiece, The Best Hollywood film o...  Positive
3      Review 4  Joaquin Phoenix gives a tour de force performa...   Neutral
4      Review 5  Most of the time movies are anticipated like t...  Negative
..          ...                                                ...       ...
103  Review 104  Truly a masterpiece, The Best Hollywood film o...  Positive
104  Review 105  The acting, cinematography, sound design and t...  Positive
105  Review 106  If you plan on watching this I would suggest s...  Negative
106  Review 107  If it wasn't for the title and make-up, you wo...  Positive
107  Review 108  Wow, what a movie! I have to admit, When I fir...  Positive

[108 rows x 3 columns]

In [7]:
# Special characters removal
df['After noise removal'] = df['Cleaned Data'].apply(lambda x: ''.join(re.sub(r"[^a-zA-Z0-9]+", ' ', charctr) for charctr in x ))
# Punctuation removal
df['Punctuation removal'] = df['After noise removal'].str.replace('[^\w\s]','')
# Remove numbers
df['Remove numbers'] = df['Punctuation removal'].str.replace('\d+', '')
# Stopwords removal
stop_word = stopwords.words('english')
df['Stopwords removal'] = df['Remove numbers'].apply(lambda x: " ".join(x for x in x.split() if x not in stop_word))
# Lower Casing
df['Lower casing'] = df['Stopwords removal'].apply(lambda x: " ".join(x.lower() for x in x.split()))
# Lemmatization
df['Lemmatization'] = df['Lower casing'].apply(lambda x: " ".join([Word(word).lemmatize() for word in x.split()]))
df

Document_ID  ...                                      Lemmatization
0      Review 1  ...  every movie come truly make impact joaquin per...
1      Review 2  ...  this movie felt alone isolated truly relate yo...
2      Review 3  ...  truly masterpiece the best hollywood film one ...
3      Review 4  ...  joaquin phoenix give tour de force performance...
4      Review 5  ...  most time movie anticipated like end falling s...
..          ...  ...                                                ...
103  Review 104  ...  truly masterpiece the best hollywood film one ...
104  Review 105  ...  the acting cinematography sound design script ...
105  Review 106  ...  if plan watching i would suggest skipping last...
106  Review 107  ...  if title make know based comic book character ...
107  Review 108  ...  wow movie i admit when i first heard joaquin p...

[108 rows x 9 columns]

In [8]:
#Extarcting words related to Corpus
from textblob import TextBlob
from collections import Counter
sentiment_text = []
for i in df['Lemmatization']:
  for term in i.split(' '):
    if TextBlob(term).sentiment.polarity != 0:
      sentiment_text.append(term)
print(sentiment_text)
word_count = Counter(sentiment_text)
word_df = pd.DataFrame(list(word_count.most_common()), columns=['Terms', 'Freq'])
word_df.index = list(range(1, len(word_count.most_common())+1))
print(word_df)

['grotesque', 'hard', 'mesmerizing', 'tragic', 'seriously', 'funny', 'far', 'typical', 'action', 'predictable', 'super', 'single', 'best', 'sorry', 'better', 'best', 'best', 'best', 'comic', 'chillingly', 'real', 'remarkable', 'surprised', 'disturbing', 'violent', 'rich', 'talented', 'exaggerated', 'believable', 'much', 'bitter', 'first', 'perfect', 'stunning', 'impossible', 'winning', 'dark', 'widely', 'live', 'action', 'many', 'impossible', 'dark', 'most', 'first', 'happy', 'complaint', 'least', 'comic', 'real', 'better', 'criminal', 'dark', 'dark', 'love', 'bad', 'amazing', 'best', 'breathtaking', 'uncomfortable', 'horrible', 'great', 'much', 'honest', 'violent', 'disturbing', 'honestly', 'expected', 'violent', 'great', 'best', 'hate', 'enjoy', 'amazing', 'enjoy', 'early', 'ill', 'trouble', 'near', 'worth', 'far', 'mere', 'main', 'slowly', 'wonderful', 'surely', 'least', 'good', 'main', 'insane', 'insane', 'much', 'superior', 'dark', 'easily', 'good', 'least', 'small', 'distinct', '

## (2) (10 points) Compare the performance of the following tools in sentiment identification: TextBlob (https://textblob.readthedocs.io/en/dev/), VADER (https://github.com/cjhutto/vaderSentiment), TFIDF-based Support Vector Machine (SVM) (Split your data into training and testing data). Take your own annotation as the standard answers. 

Reference code: https://towardsdatascience.com/fine-grained-sentiment-analysis-in-python-part-1-2697bb111ed4

In [12]:
# Write your code here
def get_sentiments_counts(sentiments):
  Negative_value = Positive_value = Neutral_value = 0
  for polarity in sentiments:
    if polarity < 0:
      Negative_value = Negative_value + 1
    elif polarity > 0:
      Positive_value = Positive_value + 1
    elif polarity == 0:
      Neutral_value = Neutral_value + 1
  return Negative_value, Positive_value, Neutral_value


# Your analysis here

 

Sentiment Analysis using Text Blob

In [13]:
nltk.download('vader_lexicon')
from textblob import TextBlob
sentiments = []
for i in df['Lemmatization']:
  sentiments.append(TextBlob(i).sentiment.polarity)
Negative_value, Positive_value, Neutral_value = get_sentiments_counts(sentiments)
print('Number of Positives - {0}'.format(Positive_value))
print('Number of Negatives - {0}'.format(Negative_value))
print('Number of Neutrals - {0}'.format(Neutral_value))

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
Number of Positives - 86
Number of Negatives - 18
Number of Neutrals - 4


VADER


In [14]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer
vader_analyzer = SentimentIntensityAnalyzer()
sentiments = []
for i in df['Lemmatization']:
  sentiments.append(vader_analyzer.polarity_scores(i)['compound'])
Negative_value, Positive_value, Neutral_value = get_sentiments_counts(sentiments)
print('Number of Positives - {0}'.format(Positive_value))
print('Number of Negatives - {0}'.format(Negative_value))
print('Number of Neutrals - {0}'.format(Neutral_value))

Number of Positives - 78
Number of Negatives - 29
Number of Neutrals - 1


/usr/local/lib/python3.7/dist-packages/nltk/twitter/__init__.py:20: UserWarning: The twython library has not been installed. Some functionality from the twitter package will not be available.
  warnings.warn("The twython library has not been installed. "


Using SVM

In [17]:

from sklearn.model_selection import train_test_split
def split_data(x_values, y_values):
  return train_test_split(x_values, y_values, test_size=0.33)

from sklearn.feature_extraction.text import TfidfVectorizer
def get_vector():
  return TfidfVectorizer(min_df = 5, max_df = 0.8, sublinear_tf = True, use_idf = True)
TfidfVectorizer = get_vector()
training_x, testing_x, training_y, testing_y = split_data(df['Lemmatization'], df['Sentiment'])
train_vectors = TfidfVectorizer.fit_transform(training_x)
test_vectors = TfidfVectorizer.transform(testing_x)
from sklearn.metrics import classification_report
from sklearn import svm
def train_predict_model(train_vectors, test_vectors, training_y):
  model = svm.SVC(kernel='linear')
  model.fit(train_vectors, training_y)
  return model.predict(test_vectors)
svm_report = classification_report(testing_y, train_predict_model(train_vectors, test_vectors, training_y), output_dict=True)
print('positive Report - ', svm_report['Positive'])
print('Negative Report - ', svm_report['Negative'])
print("Neutral Report - ", svm_report['Neutral'])

positive Report -  {'precision': 0.6551724137931034, 'recall': 0.8636363636363636, 'f1-score': 0.7450980392156864, 'support': 22}
Negative Report -  {'precision': 0.5714285714285714, 'recall': 0.3076923076923077, 'f1-score': 0.4, 'support': 13}
Neutral Report -  {'precision': 0.0, 'recall': 0.0, 'f1-score': 0.0, 'support': 1}


/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [18]:

def get_actual_sentiment_counts(sentiments):
  Negative_value = Positive_value = Neutral_value = 0
  for polarity in sentiments:
    if polarity == 'Negative':
      Negative_value = Negative_value + 1
    elif polarity == 'Positive':
      Positive_value = Positive_value + 1
    elif polarity == 'Neutral':
      Neutral_value = Neutral_value + 1
  return Negative_value, Positive_value, Neutral_value

In [20]:
Actual_Negative_Value, Actual_Positive_Value, Actual_Neutral_Value = get_actual_sentiment_counts(df['Sentiment'])

print('Number of Actual Positives - {0}'.format(Actual_Positive_Value))
print('Number of Actual Negatives - {0}'.format(Actual_Negative_Value))
print('Number of Actual Neutrals - {0}'.format(Actual_Neutral_Value))

Number of Actual Positives - 61
Number of Actual Negatives - 34
Number of Actual Neutrals - 13


Overall Results:
Number of Actual Positives - 61

Number of Actual Negatives - 34

Number of Actual Neutrals - 1


positive Report -  {'precision': 0.6551724137931034, 'recall': 0.8636363636363636, 'f1-score': 0.7450980392156864, 'support': 22}
Negative Report -  {'precision': 0.5714285714285714, 'recall': 0.3076923076923077, 'f1-score': 0.4, 'support': 13}
Neutral Report -  {'precision': 0.0, 'recall': 0.0, 'f1-score': 0.0, 'support': 1}

VADER:
Number of Positives - 78 

Number of Negatives - 29

Number of Neutrals - 1

TEXT BLOB:
Number of Positives - 86

Number of Negatives - 18

Number of Neutrals - 4

From above results , we can say that SVM has better accuracy of polarities when compared to TextBlob and Vader and among TextBlob and Vader, Textblob is better